[lsp] Basic "language server protocol" server.

ejgallego opened the pull request:
This is a first version on a language protocol server for LambdaPI
developed in the context of the internship of Ismail Lacheb by him,
Frédéric Blanqui, Emilio J. Gallego Arias, and Rodolphe Lepigre.

The server has been tested with Atom and Emacs, and supports the basic
operations. In previous versions it also supported an extension to
send proof goals as a diagnostic, but this is now currently disabled
due to upstream API issues.

Some notes:

- Lines in LSP start at 0 vs Dedukti where they start at 1.
- In addition to the base basic LSP protocol, we support
  the `documentSymbol` request.
- The server has an option to limit output to a standard LSP subset,
  see --help. This is needed by Emacs' eglot, which otherwise chokes.

ejgallego commented:
@rlepigre Note the two changes in `src`, let me know if you are OK with them.

Don't bother looking at the rest yet, I will fully replace the document manager, the current code is is just a placeholder I ported from "Demon" so we can test the connection for now.

ejgallego commented on src/parser.ml:
```diff
@@ -302,12 +302,14 @@ let blank buf pos =
   in
   fn `Ini [] (buf, pos) (buf, pos)
 
+let cmdlist_parser : p_cmd loc list Earley.grammar = (parser {l:cmd "."}*)
```
Sure, please adapt as you better see fit and I'll rebase, thanks @rlepigre .

> Do you want me to provide a parsing function taking a string as input (instead of a file) as well?

I think I only need the proper `'a grammar` values, the document manager usually wants to implement some form of incremental parsing so it will use some lower-level `Earley` itself.

Note also that I had to make `command` transparent in `Pure`, otherwise I see no way of using it.

I will also submit an issue w.r.t. the interface to `Sign` and `Pure`. Cheers.

rlepigre replied:
OK, I applied the changes you need to the `pa` branch. I'll let you merge.

ejgallego commented:
Rebased, working fine for now.

ejgallego commented:
@rlepigre I still have to handle `Sign.loading` and `loaded` imperatively; this is not a problem but just mentioning it here in case you want to refine `initial_state` to a function that creates and registers a new signature.

Indeed, in the very long-term roadmap, interactive use were several signatures can be simultaneously current is not such an uncommon use-case.

rlepigre commented:
Could you move the PR to "master"? I want to delete the "pa" branch.

rlepigre commented:
@ejgallego I think that I fixed the pure interface in the `pureinterface` branch, using a new version of the `Timed` module. Can you tell me how I can test it?

ejgallego commented:
> @ejgallego I think that I fixed the pure interface in the pureinterface branch, using a new version of the Timed module. Can you tell me how I can test it?

I just rebased this branch on top of the `pureinterface` one, but it seems things are more broken than before.

Using `master` works a bit better; also, I wonder if having a branch for this is overkill, maybe this kind of stuff should directly go to master.

rlepigre commented:
I'm not too surprised that it does not work, because I was unable to test what I done. Can you tell me how I can reproduce the problems?

ejgallego commented:
@rlepigre ah sorry I was under the impression you had a working atom setup as Isma was around. You need to:
- checkout and build this branch `jbuilder build @install`
- install atom and the dedukti-editor atom package https://github.com/lachhebo/dedukti-editor
- edit the `config.json` file so it points to the newly built server, in my case I have: 
```
      "lspServerPath": {
        "type": "string",
        "default": "/home/egallego/research/lambdapi/_build/install/default/bin/lp-lsp"
      },
```

rlepigre commented:
Note that I don't want to merge `pureinterface` into `master` as the former may introduce performance / memory usage issues. I want to do proper benchmarking before merging, but this will be done soon I think.

ejgallego commented:
Great!! Thanks @rlepigre ! I will rebase and push a new OPAM package ASAP.

ejgallego commented:
I dunno if we are talking about the same thing, but sometimes is desirable to run in batch mode without overhead indeed. What is done for example in Coq is to only snapshot under demand, not at assignement time.

craff commented:
On 18-07-03 01:21:26, Emilio Jesús Gallego Arias wrote:
> I dunno if we are talking about the same thing, but sometimes is desirable to
> run in batch mode without overhead indeed. What is done for example in Coq is
> to only snapshot under demand, not at assignement time.


Hello,

with Timed (especially the new "compact" branch), there would be no overhead.
A snashot is taken only if you call Time.save (). The (:=) is just a bit slower, but still O(1).

Cheers,
Christophe

> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

ejgallego commented:
Ok folks, as discussed I am removing the `dev` version and disabling the generation of OPAM packages from this branch.

I guess I shall close this PR as lp-lsp is not intended to be merged in the main repository; my own repos will likely contain a copy of LP, however packages will be [like today] will be built against whatever LP is in the system.

This will make development much more painful tho in the sense that pushing changes to LP will be way less agile.

ejgallego commented:
Ok, I've squashed all the commits and pushed to this branch. IMO this could be ready for merge. However note that there is a bug in the OCaml code, maybe due to a state problem?

This code used to work:
```ocaml
  let sym = Pure.in_state final_st (fun () -> !(Sign.(current_sign ()).symbols)) () in
  let sym = Extra.StrMap.fold (fun _ (s,p) l ->
      let open Terms in
      (* LIO.log_error "sym" (s.sym_name ^ " | " ^ Format.asprintf "%a" pp_term !(s.sym_type)); *)
      option_cata (fun p -> mk_syminfo file
                      (s.sym_name, s.sym_path, kind_of_type s, p) :: l) p l) sym [] in
```
but now we get a `Not_found`. I dunno what has changed.

ejgallego commented:
Another point, the code requires OCaml 4.05.0 or later, as witnessed by Travis. Should I adapt it 4.04.x?

rlepigre commented:
OK, if this is only about `List.assoc_opt`, then I think that it is better to adapt it. Note that you can rely on `Extra` (`src/extra.ml`), which extends several modules of the standard library.

If this requires more work than that, don't bother an remove support for `4.04`.

ejgallego commented:
4.04 should work now.

Still wondering about the problem with retrieving symbols for that particular state. That may need more investigation as it makes the server misbehave, or we can just disable the feature for now.

rlepigre commented:
@ejgallego I don't follows what you mean by retrieving the symbols. Can you remind me?

ejgallego commented:
> @ejgallego I don't follows what you mean by retrieving the symbols. Can you remind me?

We used this to generate the list of symbols lambdapi knows about:
```
  let sym = Pure.in_state final_st (fun () -> !(Sign.(current_sign ()).symbols)) () in
```
For some reason it is not working anymore. I will research more but maybe something rings off of your head?

rlepigre commented:
As far as I'm concerned, that should still be working. Do you get a specific error?

ejgallego commented:
> As far as I'm concerned, that should still be working. Do you get a specific error?

I got bitten by a stale bin from some old package, problem is solved I think [the server had a bug]

Should be ready to merge I think. We should add some basic tests to check the server is working tho.

rlepigre commented:
OK cool, thanks! I'll merge when Travis is done, and I'll write a `New_pure` module that you can use for the new syntax.

ejgallego commented:
Thanks.

rlepigre commented:
The `New_pure` module is in the repository, it has the same interface as `Pure`, so migration only require `module Pure = New_pure` in the files `lp-lsp/lp_lsp.ml` and `lp-lsp/lp_doc.ml`. I testes that, but dit not commit since I'm not sure how to test (I'll stop for today). Note that you can find `.lp` files in the `proofs` and `example` folders.

ejgallego commented:
Cool, will have a look ASAP.
